### PR TITLE
Added ANTLRv4 as a dependency in preparation for a full UDUNITS-2 grammar for parsing.

### PR DESCRIPTION
**Pull request by pelson**
This PR is the first of 2 to add a capability to parse UDUNITS-2 using a formal grammar (which can [hopefully be upstreamed](https://github.com/Unidata/UDUNITS-2/issues/81) to the UDUNITS-2 documentation).

This PR *does not add the grammar, nor the units AST*. It is entirely about adding the infrastructure needed to handle the formal grammar.

It means that we will be now depending (optionally) upon the ANTLR runtime (which is a pure python library with no dependencies). The most sensible runtime is *python 3 only*. I therefore propose that this capability becomes **python 3 only**.  Rather than dragging the entire package into being py3 only (a thing we need to do at *some* point this year), I've added the appropriate changes to cover things such as:

 * skip certain py3 grammar files when running pytest on python 2 (else you get SyntaxErrors!)
 * update flake8 and sticker to use python 3
 * document the compile step to turn the formal grammar into code

### CODE REVIEW

File: cf_units/__init__.py [hunk 1]
```diff
@@ -1471,7 +1471,7 @@ def __str__(self):
         __unicode__ = __str__
 
         def __str__(self):
-            return unicode(self).encode('utf8')
+            return unicode(self).encode('utf8')  # noqa: F821
```
Comment (pelson) [inline cf_units/__init__.py:1487]: Object (unicode) not recognised in python 3 😉.

File: cf_units/_udunits2_parser/compile.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+
+JAR_NAME = 'antlr-4.7.2-complete.jar'
+JAR_URL = f'https://www.antlr.org/download/{JAR_NAME}'
+HERE = Path(__file__).resolve().parent
+
+JAR = HERE / JAR_NAME
+
+LEXER = HERE / 'parser' / 'udunits2Lexer.g4'
+PARSER = Path('udunits2Parser.g4')
```
Comment (pelson) [inline cf_units/_udunits2_parser/compile.py]: HERE /

**Comment by pelson** [comment]
Looks like the sticker failure needs addressing - works fine in #139.

**Comment by pelson** [comment]
🎉 - hopefully you can see the wood from the trees @bjlittle 😉

File: cf_units/_udunits2_parser/README.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+changes to the grammar being proposed so that the two can remain in synch.
+
+### Testing the gammar
+
+An extensive set of tests exist to confirm that the parser produces equivalent results
+to the UDUNITS-2 reference implementation.
+The usual ``pytest cf_units`` includes these tests by default.
+They are typically very fast to execute, but significant slowdowns were experienced
+(x15 slowdown) when left-recursive rules were not carefully implemented, therefore please
+be consious of performance when modifying the grammar.
```
Comment (pelson) [inline cf_units/_udunits2_parser/README.md]: Typo

File: cf_units/conftest.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+if six.PY2:
+    here = os.path.dirname(__file__)
+    all_parse_py = glob.glob(
+        os.path.join(here, '_udunits2_parser', '*', '*.py'))
+    all_compiled_parse_py = glob.glob(
+        os.path.join(here, '_udunits2_parser', '*.py'))
+    collect_ignore = list(all_parse_py) + list(all_compiled_parse_py)
+
+    print(collect_ignore)
```
Comment (bjlittle) [inline cf_units/conftest.py]: @pelson Remove the `print` and add comments please

File: cf_units/_udunits2_parser/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+This sub-package is an attempt to re-implement the
+[UDUNITS-2 grammar](https://www.unidata.ucar.edu/software/udunits/udunits-2.0.4/udunits2lib.html#Grammar)
+in an implementation agnostic manner, using the power of ANTLRv4.
+
+The [lexer](udunits2Lexer.g4.jinja) is a Jinja2 template which expands
+the multi-mode tokenizer of the UDUNITS-2 grammar.
+The [parser](udunits2Parser.g4) defines the grammar of the tokens, along with
+a number of convenient lexical elements.
+
+
```
Comment (bjlittle) [inline cf_units/_udunits2_parser/README.md:12]: @pelson Nuke the extra blank line...

File: cf_units/_udunits2_parser/compile.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    echo udunits2*.g4* | entr -c "python compile.py"
+
+You're welcome ;).
+
+"""
+
+import collections
+from pathlib import Path
+import subprocess
+import re
```
Comment (bjlittle) [inline cf_units/_udunits2_parser/compile.py:32]: @pelson trivial import order change - move `re` above `subprocess`

File: setup.cfg [hunk 7]
```diff
@@ -28,4 +29,6 @@ addopts =
     --cov=cf_units
     --cov-report term-missing
     --doctest-modules
+
+
```
Comment (bjlittle) [inline setup.cfg]: @pelson minor - remove additional blank lines

**Comment by bjlittle** [comment]
@pelson A couple of super trivial changes and we're good to go!

File: cf_units/_udunits2_parser/README.md (hunk 5 above)
Comment (pelson) [reply cf_units/_udunits2_parser/README.md:12]: Did the opposite and added some more blank lines to other paragraphs (improves readability)

**Comment by bjlittle** [comment]
@pelson Thanks! Once travis-ci is happy, then so am I :+1:

**Comment by pelson** [comment]
Thanks @bjlittle. PR incoming with the actual parser... 🎉

**Pull request merged**